src/speech_recognition/hmi_server_kaldi_client.py

```diff
@@ -0,0 +1,40 @@
+import rospy
+from hmi import AbstractHMIServer, HMIResult
+
+# New imports
+from speech_recognition.kaldi_gstreamer_app import KaldiGstApp
+import sys
+import os
+# Gstreamer imports
+import gi
+gi.require_version('Gst', '1.0')
+from gi.repository import GObject, Gst
+
+
+class HMIServerKaldiClient(AbstractHMIServer):
+
+    def __init__(self):
+        """
+        DragonflyHMIServer that exposes the HMI ROS Server and holds a socket client that talks to the dragonfly
+        speech recognition server.
+        """
+        super(HMIServerKaldiClient, self).__init__(rospy.get_name())
+        self.model_path = rospy.get_param("/kaldi_model_path")     # Assuming this param has been set somewhere
+
+    def _determine_answer(self, description, grammar, target, is_preempt_requested):
+        # Todo: Take in grammar and target and compose the HCLG.fst (speech recognition graph) through a different
+        #   class or whatever, before initialising the KaldiGstApp
+
+        kaldi_app = KaldiGstApp(self.model_path)
+
+        while not kaldi_app.sentence and not rospy.is_shutdown():
+            print("No sentence received so far...")
+            rospy.sleep(1)
```

Comment:
If you want to use a sleep, maybe don't wait for a full second but maybe a 0.1s. OR, maybe even better; use a callback to let KaldiGstApp let HMIServerKaldiClient know there is a sentence ready.

Reply:
Also as I have not tested this yet, I am not sure about the reason of having a sleep command here. I'll do this tomorrow night as I need to test the killing of the gstreamer pipeline too.

Reply:
From what i gather from the code, there is a background process that sets `kaldi_app.sentence`. If that is not yet set, the background process is not yet done. You can check it without the sleep, but then would run a very busy loop until there is a sentence. Having s shorter sleep isn't nice, but a middle ground between this kind of polling and getting a callback (a notification)